### ISSUE TITLE
OpenSSL server is 0-RTT intolerant by default

### ISSUE THREAD

**Issue by tomato42** [openssl/openssl#6647]
When server is configured to support TLS 1.3 (358ffa05cd3a0 default), it will not handle a 0-RTT resume by a client gracefully - the connection will be aborted after server Finished with an `unexpected_message` alert.

this is in contrast to the draft-28 that server [MUST](https://tools.ietf.org/html/draft-ietf-tls-tls13-28#page-58) either:
 * "Ignore the extension and return a regular 1-RTT response."...
 * "Request that the client send another ClientHello by responding with a HelloRetryRequest."...
 * "Return its own "early_data" extension in EncryptedExtensions"...

Not following that requirement will cause the server to become intolerant to 0-RTT handshakes, after the 0-RTT support was disabled.

Note: here I mean just the ability to tolerate an 0-RTT handshake, not actually using the encrypted data.

Note: there are `-max_early_data` and `-early_data` switches to s_server, but I don't think enabling 0-RTT _support_ should be required to make implementation 0-RTT _tolerant_

   <details><summary>reproducer</summary><p>

```
openssl req -x509 -newkey rsa -keyout localhost.key -out localhost.crt -subj /CN=localhost -nodes -batch
openssl s_server -key localhost.key -cert localhost.crt -www 2>server.err >server.out &
openssl_pid=$!
git clone https://github.com/tomato42/tlsfuzzer
pushd tlsfuzzer
# won't be needed after tlsfuzzer#423 is merged
git checkout 0rtt-garbage-resumption
git clone https://github.com/tomato42/tlslite-ng .tlslite-ng
ln -s .tlslite-ng/tlslite tlslite
git clone https://github.com/warner/python-ecdsa .python-ecdsa
ln -s .python-ecdsa/ecdsa ecdsa
PYTHONPATH=. python scripts/test-tls13-0rtt-garbage.py
popd
kill $openssl_pid
```

The script supports a `--num-bytes` option allowing the script to change the amount of fake data sent as early data.

   </p></details>

   <details><summary>tlsfuzzer output</summary><p>

```
sanity ...
OK

handshake with invalid PSK ...
OK

handshake with invalid 0-RTT ...
Error encountered while processing node <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f3a48d7a550> (child: <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f3a48d7a550>) with last message being: <tlslite.messages.Message object at 0x7f3a48d7acd0>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-0rtt-garbage.py", line 262, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, unexpected_message)

sanity ...
OK

Basic check if TLS 1.3 server can handle 0-RTT handshake
Verify that the server can handle a 0-RTT handshake from client
even if (or rather, especially if) it doesn't support 0-RTT.

version: 1

Test end
successful: 3
failed: 1
  'handshake with invalid 0-RTT'
```

   </p></details>

   <details><summary>openssl output</summary><p>

```
Using default temp DH parameters
ACCEPT
140180740319040:error:142140A4:SSL routines:early_data_count_ok:too much early data:ssl/record/ssl3_record.c:129:
```

   </p></details>

**Comment by mattcaswell** [comment]
The problem is that the `max_early_data` setting controls both how much early_data we are willing to accept/skip over, *and* the amount of early_data we advertise as supported in our session tickets. So, changing the config from some positive value to 0 will mean we stop advertising early data support in our tickets and also immediately mean we will fail on any incoming early data as a result of tickets issued earlier.

Perhaps there needs to be *2* controls: one for how much early data support we advertise (default 0) and one for how much early data we are willing to skip over (default some largish positive integer).

**Comment by richsalz** [comment]
I am not sure this is wrong behavior.  The server shouldn't silently ignore early data, because then the client and server will disagree about what was sent.  Or is that not the issue herE?

**Comment by tomato42** [comment]
when the server does not include early_data in Encrypted Extensions than the client will know that the early data it sent was rejected

so no, they will agree on what was and wasn't sent (also, it's application_data, so handshake hashes don't come into play)

**Comment by kroeckx** [comment]
The server will indicate that it accepted it or not, see SSL_get_early_data_status()

**Comment by kaduk** [comment]
It does seem that any sort of graceful transition would basically require separate configuration for the value that goes in the ticket and the value that is used when accepting a handshake.  In the absence of that, one would send ticket_lifetime of zero (or just not send tickets) for long enough that the 0-RTT-capable tickets would age out, and only then change the max early data size.  But losing all resumption around graceful change of max early data size is a really heavy hammer, so separate knobs seems like the way to go.  (As painful as it is to keep adding more and more state to the SSL/SSL_CTX structures.)

**Comment by kroeckx** [comment]
Is there a reason not to discard all of it when it's set to 0?

**Comment by mattcaswell** [comment]
> Is there a reason not to discard all of it when it's set to 0?

The whole meaning of max_early_data is to control how much we will discard before we abort the connection. If it is set to 0 then, by definition, it means don't discard any just abort on early_data.

The problem is the same variable is *also* used to control how much early_data we advertise as being supported in our session tickets.

**Comment by kroeckx** [comment]
What happens if it's changed to a lower value (but not 0), and you
then receive something between the new and the old limit?

**Comment by kaduk** [comment]
> What happens if it's changed to a lower value (but not 0), and you then receive something between the new and the old limit?

With the current code, my understanding is that the server breaks its word and aborts the connection.
(To be clear, I think this is a bad state to be in and we should change the code.)

**Comment by mattcaswell** [comment]
Fix in #6655.
